### PR TITLE
[WIP - DO NOT MERGE] Remove type checking hacks

### PR DESCRIPTION
**Pull request by rudkx**
Experimentation with removing some performance hacks to see how much impact they have on compile times.

### CODE REVIEW

**Comment by rudkx** [comment]
@swift-ci Please smoke test

File: lib/Sema/CSSolver.cpp [hunk 1]
```diff
@@ -1645,7 +1645,7 @@ bool ConstraintSystem::solveRec(SmallVectorImpl<Solution> &solutions,
 static bool shortCircuitDisjunctionAt(Constraint *constraint,
                                       Constraint *successfulConstraint,
                                       ASTContext &ctx) {
-
+  return false;
```
Comment (slavapestov) [inline lib/Sema/CSSolver.cpp]: How can this have an effect on SIL optimizer time? Are we picking generic overloads where previously we picked non-generic ones, and now we're doing more work in inlining/specialization?

Comment (rudkx) [inline lib/Sema/CSSolver.cpp]: I haven't been able to reproduce that slow down again, but if I could...I would have no idea why.

**Comment by rudkx** [comment]
@swift-ci Please smoke test

**Comment by xedin** [comment]
@swift-ci please smoke test OS X Platform

**Pull request closed without merging**